iroh-gossip/src/proto/state.rs

```diff
@@ -0,0 +1,253 @@
+    PeerDisconnected(PA),
+    /// Update the opaque peer data about yourself.
+    UpdatePeerData(PeerData),
+}
+
+#[derive(Debug, Clone)]
+pub enum OutEvent<PA> {
+    /// Send a message on the network
+    SendMessage(PA, Message<PA>),
+    /// Emit an event to the application.
+    EmitEvent(TopicId, topic::Event<PA>),
+    /// Schedule a timer. The runtime is responsible for sending an [InEvent::TimerExpired]
+    /// after the duration.
+    ScheduleTimer(Duration, Timer<PA>),
+    /// Close the connection to a peer on the network level.
+    DisconnectPeer(PA),
+    /// Updated peer data
+    PeerData(PA, PeerData),
+}
+
+type ConnsMap<PA> = HashMap<PA, HashSet<TopicId>>;
+type Outbox<PA> = Vec<OutEvent<PA>>;
+
+enum InEventMapped<PA> {
+    All(topic::InEvent<PA>),
+    TopicEvent(TopicId, topic::InEvent<PA>),
+}
+
+impl<PA> From<InEvent<PA>> for InEventMapped<PA> {
+    fn from(event: InEvent<PA>) -> InEventMapped<PA> {
+        match event {
+            InEvent::RecvMessage(from, Message { topic, message }) => {
+                Self::TopicEvent(topic, topic::InEvent::RecvMessage(from, message))
+            }
+            InEvent::Command(topic, command) => {
+                Self::TopicEvent(topic, topic::InEvent::Command(command))
+            }
+            InEvent::TimerExpired(Timer { topic, timer }) => {
+                Self::TopicEvent(topic, topic::InEvent::TimerExpired(timer))
+            }
+            InEvent::PeerDisconnected(peer) => Self::All(topic::InEvent::PeerDisconnected(peer)),
+            InEvent::UpdatePeerData(data) => Self::All(topic::InEvent::UpdatePeerData(data)),
+        }
+    }
+}
+
+pub struct State<PA, R> {
+    me: PA,
+    me_data: PeerData,
+    config: Config,
+    rng: R,
+    states: HashMap<TopicId, topic::State<PA, R>>,
+    outbox: Outbox<PA>,
+    conns: ConnsMap<PA>,
+}
+
+impl<PA: PeerAddress, R: Rng + Clone> State<PA, R> {
+    pub fn new(me: PA, me_data: PeerData, config: Config, rng: R) -> Self {
+        Self {
+            me,
+            me_data,
+            config,
+            rng,
+            states: Default::default(),
+            outbox: Default::default(),
+            conns: Default::default(),
+        }
+    }
+
+    pub fn endpoint(&self) -> &PA {
+        &self.me
+    }
+
+    pub fn state(&self, topic: &TopicId) -> Option<&topic::State<PA, R>> {
+        self.states.get(topic)
+    }
+
+    pub fn state_mut(&mut self, topic: &TopicId) -> Option<&mut topic::State<PA, R>> {
+        self.states.get_mut(topic)
+    }
+
+    pub fn has_active_peers(&self, topic: &TopicId) -> bool {
+        self.state(topic)
+            .map(|s| s.has_active_peers())
+            .unwrap_or(false)
+    }
+
+    pub fn handle(
+        &mut self,
+        event: InEvent<PA>,
+        now: Instant,
+    ) -> impl Iterator<Item = OutEvent<PA>> + '_ {
+        let event: InEventMapped<PA> = event.into();
+
+        // todo: add command to leave a topic
```

Comment:
we should probably do this?

Reply:
Yes, good point. I added a `Quit` command.